File: bridge/tests/relay-slot.test.ts
// Mirror of `packages/antgrid_relay_client/test/relay_slot_test.dart`. The two
// implementations are hand-written copies, so a silent divergence here is an
// admission failure: the bridge would look the phone up under an id no store
// holds. Keep the cases in lockstep.
import { test, expect } from "bun:test";
import { baseSlotDeviceId, slotMachineDeviceId } from "../src/relay-slot";

test("a slot resolves back to the account device identity is keyed by", () => {
  expect(baseSlotDeviceId("device-1#machine-a")).toBe("device-1");
  expect(slotMachineDeviceId("device-1#machine-a")).toBe("machine-a");
});

test("an unscoped id passes through unchanged and claims no machine", () => {
  // Every pre-slot client sends one, and `isForeignSlot` must not reject it.
  expect(baseSlotDeviceId("device-1")).toBe("device-1");
  expect(slotMachineDeviceId("device-1")).toBeNull();
});

test("a machine id containing the separator does not corrupt the base", () => {
  // First separator wins, matching the Dart side; anything after it is scope.
  expect(baseSlotDeviceId("device-1#machine-a#extra")).toBe("device-1");
  expect(slotMachineDeviceId("device-1#machine-a#extra")).toBe("machine-a#extra");
});

File: packages/antgrid_relay_client/test/relay_slot_test.dart
// The app's relay slot id. Mirrored by `bridge/src/relay-slot.ts` — a
// divergence here is an admission failure on the bridge, so these cases pin the
// exact string shape both sides agree on.
import 'package:antgrid_relay_client/antgrid_relay_client.dart';
import 'package:test/test.dart';

void main() {
  test('a slot scopes the account device by the machine it dials', () {
    expect(relaySlotId('device-1', 'machine-a'), 'device-1#machine-a');
  });

  test('every machine gets a distinct slot for one account device', () {
    // The whole point: the relay arbitrates per `hello.deviceId` and supersedes
    // an equal epoch, so two machines wanted at once must not share a slot.
    expect(
      relaySlotId('device-1', 'machine-a'),
      isNot(relaySlotId('device-1', 'machine-b')),
    );
  });

  test('a slot resolves back to the account device identity is keyed by', () {
    expect(baseSlotDeviceId(relaySlotId('device-1', 'machine-a')), 'device-1');
  });

  test('an unscoped id passes through unchanged', () {
    // A client that dials without a slot must keep resolving — the bridge runs
    // the same strip over every route id it receives, ours or not.
    expect(baseSlotDeviceId('device-1'), 'device-1');
  });

  test('a machine id containing the separator does not corrupt the base', () {
    // First separator wins on both sides; anything after it is scope.
    expect(baseSlotDeviceId('device-1#machine-a#extra'), 'device-1');
  });
}
